flang/lib/Lower/ConvertVariable.cpp

```diff
@@ -1492,6 +1494,52 @@ static void genDeclareSymbol(Fortran::lower::AbstractConverter &converter,
     auto name = converter.mangleName(sym);
     fir::FortranVariableFlagsAttr attributes =
         Fortran::lower::translateSymbolAttributes(builder.getContext(), sym);
+
+    if (isCrayPointee) {
+      mlir::Type baseType =
+          hlfir::getFortranElementOrSequenceType(base.getType());
+      if (auto seqType = mlir::dyn_cast<fir::SequenceType>(baseType)) {
+        // The pointer box's sequence type must be with unknown shape.
+        llvm::SmallVector<int64_t> shape(seqType.getDimension(),
+                                         fir::SequenceType::getUnknownExtent());
+        baseType = fir::SequenceType::get(shape, seqType.getEleTy());
+      }
+      fir::BoxType ptrBoxType =
+          fir::BoxType::get(fir::PointerType::get(baseType));
+      mlir::Value boxAlloc =
+          fir::factory::genNullBoxStorage(builder, loc, ptrBoxType);
```

Comment:
This code allocates and initialize the pointer storage, I think the initialization is not needed/confusing in the generated IR given it is done again with the actual shape info below.

Maybe you can just use builder.createTemporary(loc, ptrBoxType) to get the temp and move the initialization code you have below before the hlfir.declare. That way, at the point of the hlfir.declare, the pointer contains the right shape/length info.

Reply:
I am not sure if moving the initialization code before hlfir.declare is necessary.  Do you have an example when this may matter?

Right now the IR looks like we have a regular local pointer with the missing initialization of the pointer into an unassociated status.  Since the following initialization happens before any possible use, I would say it does not matter whether it happens before or after hlfir.declare, but I may be missing something.

Reply:
It is not necessary, this was a (very minor) personal opinion regarding the readability of the generated IR. I am also fine with the way it is.

Reply:
Oh, okay.  Thank you for the clarification!